libqtile/widget/battery_monitor.py

```diff
@@ -0,0 +1,394 @@
+from libqtile import bar
+from . import base
+from .. import images
+
+
+class State(Enum):
+    CHARGING = 0
+    DISCHARGING = 1
+    FULL = 2
+    EMPTY = 3
+    UNKNOWN = 4
+
+
+class _Battery(object):
+    """
+        Battery interface class specifying what member functions a
+        battery implementation should provide.
+    """
+
+    def _update_data(self):
+        """
+            Reads the battery status from the system and stores the
+            result internally for later retrieval. Is called every
+            time battery status is queried.
+
+            No return
+
+            Raises RuntimeError on error.
+        """
+        raise NotImplementedError
+
+    def _get_state(self):
+        """
+            Returns the state of the battery.
+
+            Return value is one of:
+                State.CHARGING
+                State.DISCHARGING
+                State.FULL
+                State.EMPTY
+                State.UNKNOWN
+
+            Raises RuntimeError on error.
+        """
+        raise NotImplementedError
+
+    def _get_percentage(self):
+        """
+            Returns the percentage remaining in the battery in decimal
+            form.
+
+            Return value:
+                float containing remaining charge percentage in
+                decimal form
+
+            Raises RuntimeError on error.
+        """
+        raise NotImplementedError
+
+    def _get_power(self):
+        """
+            Returns the current power consumption in watts [W] as a
+            number.
+
+            Return value:
+                Number containing the current power consumption in
+                watts [W].
+
+            Raises RuntimeErorr on error.
+        """
+        raise NotImplementedError
+
+    def _get_time(self):
+        """
+            Returns the estimated remaining battery time in seconds.
+
+            Return value:
+                Number containing the seconds of battery time
+                remaining.
+
+            Raises RuntimeError on error.
+        """
+        raise NotImplementedError
+
+
+class _FreeBSDBattery(_Battery):
+    """
+        A battery class compatible with FreeBSD. Reads battery status
+        using acpiconf.
+
+        Takes a battery setting containing the number of the battery
+        that should be monitored.
+    """
+    def __init__(self, battery='0'):
+        self.battery = battery
+        self._update_data()
+
+    def _update_data(self):
+        try:
+            self.info = \
+                check_output(['acpiconf', '-i', self.battery]).decode('utf-8')
+        except CalledProcessError:
+            raise RuntimeError('acpiconf exited incorrectly')
+
+    def _get_state(self):
+        stat = re.search(r'State:\t+([a-z]+)', self.info)
+
+        if stat is None:
+            raise RuntimeError('Could not get battery state!')
+
+        stat = stat.group(1)
+        if stat == 'charging':
+            result = State.CHARGING
```

Comment:
No need to assign a variable in all of these, just return the values

Reply:
I'm of the opinion that assigning a value to a result and returning it at the end of a function is cleaner and more maintainable as it reduces code complexity. If you feel strongly about it, I can of course change it.

Reply:
No problem, I'll leave that up to preference.